// @kava-labs/javascript-sdk/lib/proto/kava/committee/v1beta1/committee.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenCommittee = exports.MemberCommittee = exports.BaseCommittee = exports.tallyOptionToJSON = exports.tallyOptionFromJSON = exports.TallyOption = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var duration_1 = require("../../../google/protobuf/duration");
var any_1 = require("../../../google/protobuf/any");
exports.protobufPackage = 'kava.committee.v1beta1';
/** TallyOption enumerates the valid types of a tally. */
var TallyOption;
(function (TallyOption) {
    /** TALLY_OPTION_UNSPECIFIED - TALLY_OPTION_UNSPECIFIED defines a null tally option. */
    TallyOption[TallyOption["TALLY_OPTION_UNSPECIFIED"] = 0] = "TALLY_OPTION_UNSPECIFIED";
    /** TALLY_OPTION_FIRST_PAST_THE_POST - Votes are tallied each block and the proposal passes as soon as the vote threshold is reached */
    TallyOption[TallyOption["TALLY_OPTION_FIRST_PAST_THE_POST"] = 1] = "TALLY_OPTION_FIRST_PAST_THE_POST";
    /** TALLY_OPTION_DEADLINE - Votes are tallied exactly once, when the deadline time is reached */
    TallyOption[TallyOption["TALLY_OPTION_DEADLINE"] = 2] = "TALLY_OPTION_DEADLINE";
    TallyOption[TallyOption["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(TallyOption = exports.TallyOption || (exports.TallyOption = {}));
function tallyOptionFromJSON(object) {
    switch (object) {
        case 0:
        case 'TALLY_OPTION_UNSPECIFIED':
            return TallyOption.TALLY_OPTION_UNSPECIFIED;
        case 1:
        case 'TALLY_OPTION_FIRST_PAST_THE_POST':
            return TallyOption.TALLY_OPTION_FIRST_PAST_THE_POST;
        case 2:
        case 'TALLY_OPTION_DEADLINE':
            return TallyOption.TALLY_OPTION_DEADLINE;
        case -1:
        case 'UNRECOGNIZED':
        default:
            return TallyOption.UNRECOGNIZED;
    }
}
exports.tallyOptionFromJSON = tallyOptionFromJSON;
function tallyOptionToJSON(object) {
    switch (object) {
        case TallyOption.TALLY_OPTION_UNSPECIFIED:
            return 'TALLY_OPTION_UNSPECIFIED';
        case TallyOption.TALLY_OPTION_FIRST_PAST_THE_POST:
            return 'TALLY_OPTION_FIRST_PAST_THE_POST';
        case TallyOption.TALLY_OPTION_DEADLINE:
            return 'TALLY_OPTION_DEADLINE';
        default:
            return 'UNKNOWN';
    }
}
exports.tallyOptionToJSON = tallyOptionToJSON;
function createBaseBaseCommittee() {
    return {
        id: long_1.default.UZERO,
        description: '',
        members: [],
        permissions: [],
        voteThreshold: '',
        proposalDuration: undefined,
        tallyOption: 0,
    };
}
exports.BaseCommittee = {
    encode: function (message, writer) {
        var e_1, _a, e_2, _b;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (!message.id.isZero()) {
            writer.uint32(8).uint64(message.id);
        }
        if (message.description !== '') {
            writer.uint32(18).string(message.description);
        }
        try {
            for (var _c = __values(message.members), _d = _c.next(); !_d.done; _d = _c.next()) {
                var v = _d.value;
                writer.uint32(26).bytes(v);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        try {
            for (var _e = __values(message.permissions), _f = _e.next(); !_f.done; _f = _e.next()) {
                var v = _f.value;
                any_1.Any.encode(v, writer.uint32(34).fork()).ldelim();
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
        if (message.voteThreshold !== '') {
            writer.uint32(42).string(message.voteThreshold);
        }
        if (message.proposalDuration !== undefined) {
            duration_1.Duration.encode(message.proposalDuration, writer.uint32(50).fork()).ldelim();
        }
        if (message.tallyOption !== 0) {
            writer.uint32(56).int32(message.tallyOption);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseBaseCommittee();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.uint64();
                    break;
                case 2:
                    message.description = reader.string();
                    break;
                case 3:
                    message.members.push(reader.bytes());
                    break;
                case 4:
                    message.permissions.push(any_1.Any.decode(reader, reader.uint32()));
                    break;
                case 5:
                    message.voteThreshold = reader.string();
                    break;
                case 6:
                    message.proposalDuration = duration_1.Duration.decode(reader, reader.uint32());
                    break;
                case 7:
                    message.tallyOption = reader.int32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            id: isSet(object.id) ? long_1.default.fromString(object.id) : long_1.default.UZERO,
            description: isSet(object.description) ? String(object.description) : '',
            members: Array.isArray(object === null || object === void 0 ? void 0 : object.members)
                ? object.members.map(function (e) { return bytesFromBase64(e); })
                : [],
            permissions: Array.isArray(object === null || object === void 0 ? void 0 : object.permissions)
                ? object.permissions.map(function (e) { return any_1.Any.fromJSON(e); })
                : [],
            voteThreshold: isSet(object.voteThreshold)
                ? String(object.voteThreshold)
                : '',
            proposalDuration: isSet(object.proposalDuration)
                ? duration_1.Duration.fromJSON(object.proposalDuration)
                : undefined,
            tallyOption: isSet(object.tallyOption)
                ? tallyOptionFromJSON(object.tallyOption)
                : 0,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.id !== undefined &&
            (obj.id = (message.id || long_1.default.UZERO).toString());
        message.description !== undefined &&
            (obj.description = message.description);
        if (message.members) {
            obj.members = message.members.map(function (e) {
                return base64FromBytes(e !== undefined ? e : new Uint8Array());
            });
        }
        else {
            obj.members = [];
        }
        if (message.permissions) {
            obj.permissions = message.permissions.map(function (e) {
                return e ? any_1.Any.toJSON(e) : undefined;
            });
        }
        else {
            obj.permissions = [];
        }
        message.voteThreshold !== undefined &&
            (obj.voteThreshold = message.voteThreshold);
        message.proposalDuration !== undefined &&
            (obj.proposalDuration = message.proposalDuration
                ? duration_1.Duration.toJSON(message.proposalDuration)
                : undefined);
        message.tallyOption !== undefined &&
            (obj.tallyOption = tallyOptionToJSON(message.tallyOption));
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e;
        var message = createBaseBaseCommittee();
        message.id =
            object.id !== undefined && object.id !== null
                ? long_1.default.fromValue(object.id)
                : long_1.default.UZERO;
        message.description = (_a = object.description) !== null && _a !== void 0 ? _a : '';
        message.members = ((_b = object.members) === null || _b === void 0 ? void 0 : _b.map(function (e) { return e; })) || [];
        message.permissions =
            ((_c = object.permissions) === null || _c === void 0 ? void 0 : _c.map(function (e) { return any_1.Any.fromPartial(e); })) || [];
        message.voteThreshold = (_d = object.voteThreshold) !== null && _d !== void 0 ? _d : '';
        message.proposalDuration =
            object.proposalDuration !== undefined && object.proposalDuration !== null
                ? duration_1.Duration.fromPartial(object.proposalDuration)
                : undefined;
        message.tallyOption = (_e = object.tallyOption) !== null && _e !== void 0 ? _e : 0;
        return message;
    },
};
function createBaseMemberCommittee() {
    return { baseCommittee: undefined };
}
exports.MemberCommittee = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.baseCommittee !== undefined) {
            exports.BaseCommittee.encode(message.baseCommittee, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMemberCommittee();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.baseCommittee = exports.BaseCommittee.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            baseCommittee: isSet(object.baseCommittee)
                ? exports.BaseCommittee.fromJSON(object.baseCommittee)
                : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.baseCommittee !== undefined &&
            (obj.baseCommittee = message.baseCommittee
                ? exports.BaseCommittee.toJSON(message.baseCommittee)
                : undefined);
        return obj;
    },
    fromPartial: function (object) {
        var message = createBaseMemberCommittee();
        message.baseCommittee =
            object.baseCommittee !== undefined && object.baseCommittee !== null
                ? exports.BaseCommittee.fromPartial(object.baseCommittee)
                : undefined;
        return message;
    },
};
function createBaseTokenCommittee() {
    return { baseCommittee: undefined, quorum: '', tallyDenom: '' };
}
exports.TokenCommittee = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.baseCommittee !== undefined) {
            exports.BaseCommittee.encode(message.baseCommittee, writer.uint32(10).fork()).ldelim();
        }
        if (message.quorum !== '') {
            writer.uint32(18).string(message.quorum);
        }
        if (message.tallyDenom !== '') {
            writer.uint32(26).string(message.tallyDenom);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseTokenCommittee();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.baseCommittee = exports.BaseCommittee.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.quorum = reader.string();
                    break;
                case 3:
                    message.tallyDenom = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            baseCommittee: isSet(object.baseCommittee)
                ? exports.BaseCommittee.fromJSON(object.baseCommittee)
                : undefined,
            quorum: isSet(object.quorum) ? String(object.quorum) : '',
            tallyDenom: isSet(object.tallyDenom) ? String(object.tallyDenom) : '',
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.baseCommittee !== undefined &&
            (obj.baseCommittee = message.baseCommittee
                ? exports.BaseCommittee.toJSON(message.baseCommittee)
                : undefined);
        message.quorum !== undefined && (obj.quorum = message.quorum);
        message.tallyDenom !== undefined && (obj.tallyDenom = message.tallyDenom);
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseTokenCommittee();
        message.baseCommittee =
            object.baseCommittee !== undefined && object.baseCommittee !== null
                ? exports.BaseCommittee.fromPartial(object.baseCommittee)
                : undefined;
        message.quorum = (_a = object.quorum) !== null && _a !== void 0 ? _a : '';
        message.tallyDenom = (_b = object.tallyDenom) !== null && _b !== void 0 ? _b : '';
        return message;
    },
};
var globalThis = (function () {
    if (typeof globalThis !== 'undefined')
        return globalThis;
    if (typeof self !== 'undefined')
        return self;
    if (typeof window !== 'undefined')
        return window;
    if (typeof global !== 'undefined')
        return global;
    throw 'Unable to locate global object';
})();
var atob = globalThis.atob ||
    (function (b64) { return globalThis.Buffer.from(b64, 'base64').toString('binary'); });
function bytesFromBase64(b64) {
    var bin = atob(b64);
    var arr = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
var btoa = globalThis.btoa ||
    (function (bin) { return globalThis.Buffer.from(bin, 'binary').toString('base64'); });
function base64FromBytes(arr) {
    var e_3, _a;
    var bin = [];
    try {
        for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
            var byte = arr_1_1.value;
            bin.push(String.fromCharCode(byte));
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
        }
        finally { if (e_3) throw e_3.error; }
    }
    return btoa(bin.join(''));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
