// @kava-labs/javascript-sdk/lib/proto/cosmos/crisis/v1beta1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var coin_1 = require("../../../cosmos/base/v1beta1/coin");
exports.protobufPackage = 'cosmos.crisis.v1beta1';
function createBaseGenesisState() {
    return { constantFee: undefined };
}
exports.GenesisState = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.constantFee !== undefined) {
            coin_1.Coin.encode(message.constantFee, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseGenesisState();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 3:
                    message.constantFee = coin_1.Coin.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            constantFee: isSet(object.constantFee)
                ? coin_1.Coin.fromJSON(object.constantFee)
                : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.constantFee !== undefined &&
            (obj.constantFee = message.constantFee
                ? coin_1.Coin.toJSON(message.constantFee)
                : undefined);
        return obj;
    },
    fromPartial: function (object) {
        var message = createBaseGenesisState();
        message.constantFee =
            object.constantFee !== undefined && object.constantFee !== null
                ? coin_1.Coin.fromPartial(object.constantFee)
                : undefined;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
