// @kava-labs/javascript-sdk/lib/proto/kava/auction/v1beta1/genesis.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var duration_1 = require("../../../google/protobuf/duration");
var any_1 = require("../../../google/protobuf/any");
exports.protobufPackage = 'kava.auction.v1beta1';
function createBaseGenesisState() {
    return { nextAuctionId: long_1.default.UZERO, params: undefined, auctions: [] };
}
exports.GenesisState = {
    encode: function (message, writer) {
        var e_1, _a;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (!message.nextAuctionId.isZero()) {
            writer.uint32(8).uint64(message.nextAuctionId);
        }
        if (message.params !== undefined) {
            exports.Params.encode(message.params, writer.uint32(18).fork()).ldelim();
        }
        try {
            for (var _b = __values(message.auctions), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                any_1.Any.encode(v, writer.uint32(26).fork()).ldelim();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseGenesisState();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.nextAuctionId = reader.uint64();
                    break;
                case 2:
                    message.params = exports.Params.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.auctions.push(any_1.Any.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            nextAuctionId: isSet(object.nextAuctionId)
                ? long_1.default.fromString(object.nextAuctionId)
                : long_1.default.UZERO,
            params: isSet(object.params) ? exports.Params.fromJSON(object.params) : undefined,
            auctions: Array.isArray(object === null || object === void 0 ? void 0 : object.auctions)
                ? object.auctions.map(function (e) { return any_1.Any.fromJSON(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.nextAuctionId !== undefined &&
            (obj.nextAuctionId = (message.nextAuctionId || long_1.default.UZERO).toString());
        message.params !== undefined &&
            (obj.params = message.params ? exports.Params.toJSON(message.params) : undefined);
        if (message.auctions) {
            obj.auctions = message.auctions.map(function (e) {
                return e ? any_1.Any.toJSON(e) : undefined;
            });
        }
        else {
            obj.auctions = [];
        }
        return obj;
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseGenesisState();
        message.nextAuctionId =
            object.nextAuctionId !== undefined && object.nextAuctionId !== null
                ? long_1.default.fromValue(object.nextAuctionId)
                : long_1.default.UZERO;
        message.params =
            object.params !== undefined && object.params !== null
                ? exports.Params.fromPartial(object.params)
                : undefined;
        message.auctions = ((_a = object.auctions) === null || _a === void 0 ? void 0 : _a.map(function (e) { return any_1.Any.fromPartial(e); })) || [];
        return message;
    },
};
function createBaseParams() {
    return {
        maxAuctionDuration: undefined,
        forwardBidDuration: undefined,
        reverseBidDuration: undefined,
        incrementSurplus: new Uint8Array(),
        incrementDebt: new Uint8Array(),
        incrementCollateral: new Uint8Array(),
    };
}
exports.Params = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.maxAuctionDuration !== undefined) {
            duration_1.Duration.encode(message.maxAuctionDuration, writer.uint32(10).fork()).ldelim();
        }
        if (message.forwardBidDuration !== undefined) {
            duration_1.Duration.encode(message.forwardBidDuration, writer.uint32(50).fork()).ldelim();
        }
        if (message.reverseBidDuration !== undefined) {
            duration_1.Duration.encode(message.reverseBidDuration, writer.uint32(58).fork()).ldelim();
        }
        if (message.incrementSurplus.length !== 0) {
            writer.uint32(26).bytes(message.incrementSurplus);
        }
        if (message.incrementDebt.length !== 0) {
            writer.uint32(34).bytes(message.incrementDebt);
        }
        if (message.incrementCollateral.length !== 0) {
            writer.uint32(42).bytes(message.incrementCollateral);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseParams();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.maxAuctionDuration = duration_1.Duration.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.forwardBidDuration = duration_1.Duration.decode(reader, reader.uint32());
                    break;
                case 7:
                    message.reverseBidDuration = duration_1.Duration.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.incrementSurplus = reader.bytes();
                    break;
                case 4:
                    message.incrementDebt = reader.bytes();
                    break;
                case 5:
                    message.incrementCollateral = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            maxAuctionDuration: isSet(object.maxAuctionDuration)
                ? duration_1.Duration.fromJSON(object.maxAuctionDuration)
                : undefined,
            forwardBidDuration: isSet(object.forwardBidDuration)
                ? duration_1.Duration.fromJSON(object.forwardBidDuration)
                : undefined,
            reverseBidDuration: isSet(object.reverseBidDuration)
                ? duration_1.Duration.fromJSON(object.reverseBidDuration)
                : undefined,
            incrementSurplus: isSet(object.incrementSurplus)
                ? bytesFromBase64(object.incrementSurplus)
                : new Uint8Array(),
            incrementDebt: isSet(object.incrementDebt)
                ? bytesFromBase64(object.incrementDebt)
                : new Uint8Array(),
            incrementCollateral: isSet(object.incrementCollateral)
                ? bytesFromBase64(object.incrementCollateral)
                : new Uint8Array(),
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.maxAuctionDuration !== undefined &&
            (obj.maxAuctionDuration = message.maxAuctionDuration
                ? duration_1.Duration.toJSON(message.maxAuctionDuration)
                : undefined);
        message.forwardBidDuration !== undefined &&
            (obj.forwardBidDuration = message.forwardBidDuration
                ? duration_1.Duration.toJSON(message.forwardBidDuration)
                : undefined);
        message.reverseBidDuration !== undefined &&
            (obj.reverseBidDuration = message.reverseBidDuration
                ? duration_1.Duration.toJSON(message.reverseBidDuration)
                : undefined);
        message.incrementSurplus !== undefined &&
            (obj.incrementSurplus = base64FromBytes(message.incrementSurplus !== undefined
                ? message.incrementSurplus
                : new Uint8Array()));
        message.incrementDebt !== undefined &&
            (obj.incrementDebt = base64FromBytes(message.incrementDebt !== undefined
                ? message.incrementDebt
                : new Uint8Array()));
        message.incrementCollateral !== undefined &&
            (obj.incrementCollateral = base64FromBytes(message.incrementCollateral !== undefined
                ? message.incrementCollateral
                : new Uint8Array()));
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c;
        var message = createBaseParams();
        message.maxAuctionDuration =
            object.maxAuctionDuration !== undefined &&
                object.maxAuctionDuration !== null
                ? duration_1.Duration.fromPartial(object.maxAuctionDuration)
                : undefined;
        message.forwardBidDuration =
            object.forwardBidDuration !== undefined &&
                object.forwardBidDuration !== null
                ? duration_1.Duration.fromPartial(object.forwardBidDuration)
                : undefined;
        message.reverseBidDuration =
            object.reverseBidDuration !== undefined &&
                object.reverseBidDuration !== null
                ? duration_1.Duration.fromPartial(object.reverseBidDuration)
                : undefined;
        message.incrementSurplus = (_a = object.incrementSurplus) !== null && _a !== void 0 ? _a : new Uint8Array();
        message.incrementDebt = (_b = object.incrementDebt) !== null && _b !== void 0 ? _b : new Uint8Array();
        message.incrementCollateral =
            (_c = object.incrementCollateral) !== null && _c !== void 0 ? _c : new Uint8Array();
        return message;
    },
};
var globalThis = (function () {
    if (typeof globalThis !== 'undefined')
        return globalThis;
    if (typeof self !== 'undefined')
        return self;
    if (typeof window !== 'undefined')
        return window;
    if (typeof global !== 'undefined')
        return global;
    throw 'Unable to locate global object';
})();
var atob = globalThis.atob ||
    (function (b64) { return globalThis.Buffer.from(b64, 'base64').toString('binary'); });
function bytesFromBase64(b64) {
    var bin = atob(b64);
    var arr = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
var btoa = globalThis.btoa ||
    (function (bin) { return globalThis.Buffer.from(bin, 'binary').toString('base64'); });
function base64FromBytes(arr) {
    var e_2, _a;
    var bin = [];
    try {
        for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
            var byte = arr_1_1.value;
            bin.push(String.fromCharCode(byte));
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return btoa(bin.join(''));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
