// @kava-labs/javascript-sdk/src/proto/kava/evmutil/v1beta1/conversion_pair.ts
/* eslint-disable */
import Long from 'long';
import _m0 from 'protobufjs/minimal';

export const protobufPackage = 'kava.evmutil.v1beta1';

/**
 * ConversionPair defines a Kava ERC20 address and corresponding denom that is
 * allowed to be converted between ERC20 and sdk.Coin
 */
export interface ConversionPair {
  /** ERC20 address of the token on the Kava EVM */
  kavaErc20Address: Uint8Array;
  /** Denom of the corresponding sdk.Coin */
  denom: string;
}

function createBaseConversionPair(): ConversionPair {
  return { kavaErc20Address: new Uint8Array(), denom: '' };
}

export const ConversionPair = {
  encode(
    message: ConversionPair,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (message.kavaErc20Address.length !== 0) {
      writer.uint32(10).bytes(message.kavaErc20Address);
    }
    if (message.denom !== '') {
      writer.uint32(18).string(message.denom);
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): ConversionPair {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseConversionPair();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.kavaErc20Address = reader.bytes();
          break;
        case 2:
          message.denom = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): ConversionPair {
    return {
      kavaErc20Address: isSet(object.kavaErc20Address)
        ? bytesFromBase64(object.kavaErc20Address)
        : new Uint8Array(),
      denom: isSet(object.denom) ? String(object.denom) : '',
    };
  },

  toJSON(message: ConversionPair): unknown {
    const obj: any = {};
    message.kavaErc20Address !== undefined &&
      (obj.kavaErc20Address = base64FromBytes(
        message.kavaErc20Address !== undefined
          ? message.kavaErc20Address
          : new Uint8Array()
      ));
    message.denom !== undefined && (obj.denom = message.denom);
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<ConversionPair>, I>>(
    object: I
  ): ConversionPair {
    const message = createBaseConversionPair();
    message.kavaErc20Address = object.kavaErc20Address ?? new Uint8Array();
    message.denom = object.denom ?? '';
    return message;
  },
};

declare var self: any | undefined;
declare var window: any | undefined;
declare var global: any | undefined;
var globalThis: any = (() => {
  if (typeof globalThis !== 'undefined') return globalThis;
  if (typeof self !== 'undefined') return self;
  if (typeof window !== 'undefined') return window;
  if (typeof global !== 'undefined') return global;
  throw 'Unable to locate global object';
})();

const atob: (b64: string) => string =
  globalThis.atob ||
  ((b64) => globalThis.Buffer.from(b64, 'base64').toString('binary'));
function bytesFromBase64(b64: string): Uint8Array {
  const bin = atob(b64);
  const arr = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; ++i) {
    arr[i] = bin.charCodeAt(i);
  }
  return arr;
}

const btoa: (bin: string) => string =
  globalThis.btoa ||
  ((bin) => globalThis.Buffer.from(bin, 'binary').toString('base64'));
function base64FromBytes(arr: Uint8Array): string {
  const bin: string[] = [];
  for (const byte of arr) {
    bin.push(String.fromCharCode(byte));
  }
  return btoa(bin.join(''));
}

type Builtin =
  | Date
  | Function
  | Uint8Array
  | string
  | number
  | boolean
  | undefined;

export type DeepPartial<T> = T extends Builtin
  ? T
  : T extends Long
  ? string | number | Long
  : T extends Array<infer U>
  ? Array<DeepPartial<U>>
  : T extends ReadonlyArray<infer U>
  ? ReadonlyArray<DeepPartial<U>>
  : T extends {}
  ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin
  ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & Record<
        Exclude<keyof I, KeysOfUnion<P>>,
        never
      >;

if (_m0.util.Long !== Long) {
  _m0.util.Long = Long as any;
  _m0.configure();
}

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
