// @kava-labs/javascript-sdk/lib/msg/router/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.router = void 0;
function newMsgMintDeposit(depositor, validator, amount) {
    return {
        type: 'router/MsgMintDeposit',
        value: {
            depositor: depositor,
            validator: validator,
            amount: amount,
        },
    };
}
function newMsgDelegateMintDeposit(depositor, validator, amount) {
    return {
        type: 'router/MsgDelegateMintDeposit',
        value: {
            depositor: depositor,
            validator: validator,
            amount: amount,
        },
    };
}
function newMsgWithdrawBurn(from, validator, amount) {
    return {
        type: 'router/MsgWithdrawBurn',
        value: {
            from: from,
            validator: validator,
            amount: amount,
        },
    };
}
function newMsgWithdrawBurnUndelegate(from, validator, amount) {
    return {
        type: 'router/MsgWithdrawBurnUndelegate',
        value: {
            from: from,
            validator: validator,
            amount: amount,
        },
    };
}
exports.router = {
    newMsgMintDeposit: newMsgMintDeposit,
    newMsgDelegateMintDeposit: newMsgDelegateMintDeposit,
    newMsgWithdrawBurn: newMsgWithdrawBurn,
    newMsgWithdrawBurnUndelegate: newMsgWithdrawBurnUndelegate,
};
