// @kava-labs/javascript-sdk/lib/proto/kava/evmutil/v1beta1/tx.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClientImpl = exports.MsgConvertERC20ToCoinResponse = exports.MsgConvertERC20ToCoin = exports.MsgConvertCoinToERC20Response = exports.MsgConvertCoinToERC20 = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var coin_1 = require("../../../cosmos/base/v1beta1/coin");
exports.protobufPackage = 'kava.evmutil.v1beta1';
function createBaseMsgConvertCoinToERC20() {
    return { initiator: '', receiver: '', amount: undefined };
}
exports.MsgConvertCoinToERC20 = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.initiator !== '') {
            writer.uint32(10).string(message.initiator);
        }
        if (message.receiver !== '') {
            writer.uint32(18).string(message.receiver);
        }
        if (message.amount !== undefined) {
            coin_1.Coin.encode(message.amount, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgConvertCoinToERC20();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.initiator = reader.string();
                    break;
                case 2:
                    message.receiver = reader.string();
                    break;
                case 3:
                    message.amount = coin_1.Coin.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            initiator: isSet(object.initiator) ? String(object.initiator) : '',
            receiver: isSet(object.receiver) ? String(object.receiver) : '',
            amount: isSet(object.amount) ? coin_1.Coin.fromJSON(object.amount) : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.initiator !== undefined && (obj.initiator = message.initiator);
        message.receiver !== undefined && (obj.receiver = message.receiver);
        message.amount !== undefined &&
            (obj.amount = message.amount ? coin_1.Coin.toJSON(message.amount) : undefined);
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseMsgConvertCoinToERC20();
        message.initiator = (_a = object.initiator) !== null && _a !== void 0 ? _a : '';
        message.receiver = (_b = object.receiver) !== null && _b !== void 0 ? _b : '';
        message.amount =
            object.amount !== undefined && object.amount !== null
                ? coin_1.Coin.fromPartial(object.amount)
                : undefined;
        return message;
    },
};
function createBaseMsgConvertCoinToERC20Response() {
    return {};
}
exports.MsgConvertCoinToERC20Response = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgConvertCoinToERC20Response();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    fromPartial: function (_) {
        var message = createBaseMsgConvertCoinToERC20Response();
        return message;
    },
};
function createBaseMsgConvertERC20ToCoin() {
    return { initiator: '', receiver: '', kavaErc20Address: '', amount: '' };
}
exports.MsgConvertERC20ToCoin = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.initiator !== '') {
            writer.uint32(10).string(message.initiator);
        }
        if (message.receiver !== '') {
            writer.uint32(18).string(message.receiver);
        }
        if (message.kavaErc20Address !== '') {
            writer.uint32(26).string(message.kavaErc20Address);
        }
        if (message.amount !== '') {
            writer.uint32(34).string(message.amount);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgConvertERC20ToCoin();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.initiator = reader.string();
                    break;
                case 2:
                    message.receiver = reader.string();
                    break;
                case 3:
                    message.kavaErc20Address = reader.string();
                    break;
                case 4:
                    message.amount = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            initiator: isSet(object.initiator) ? String(object.initiator) : '',
            receiver: isSet(object.receiver) ? String(object.receiver) : '',
            kavaErc20Address: isSet(object.kavaErc20Address)
                ? String(object.kavaErc20Address)
                : '',
            amount: isSet(object.amount) ? String(object.amount) : '',
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.initiator !== undefined && (obj.initiator = message.initiator);
        message.receiver !== undefined && (obj.receiver = message.receiver);
        message.kavaErc20Address !== undefined &&
            (obj.kavaErc20Address = message.kavaErc20Address);
        message.amount !== undefined && (obj.amount = message.amount);
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d;
        var message = createBaseMsgConvertERC20ToCoin();
        message.initiator = (_a = object.initiator) !== null && _a !== void 0 ? _a : '';
        message.receiver = (_b = object.receiver) !== null && _b !== void 0 ? _b : '';
        message.kavaErc20Address = (_c = object.kavaErc20Address) !== null && _c !== void 0 ? _c : '';
        message.amount = (_d = object.amount) !== null && _d !== void 0 ? _d : '';
        return message;
    },
};
function createBaseMsgConvertERC20ToCoinResponse() {
    return {};
}
exports.MsgConvertERC20ToCoinResponse = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgConvertERC20ToCoinResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    fromPartial: function (_) {
        var message = createBaseMsgConvertERC20ToCoinResponse();
        return message;
    },
};
var MsgClientImpl = /** @class */ (function () {
    function MsgClientImpl(rpc) {
        this.rpc = rpc;
        this.ConvertCoinToERC20 = this.ConvertCoinToERC20.bind(this);
        this.ConvertERC20ToCoin = this.ConvertERC20ToCoin.bind(this);
    }
    MsgClientImpl.prototype.ConvertCoinToERC20 = function (request) {
        var data = exports.MsgConvertCoinToERC20.encode(request).finish();
        var promise = this.rpc.request('kava.evmutil.v1beta1.Msg', 'ConvertCoinToERC20', data);
        return promise.then(function (data) {
            return exports.MsgConvertCoinToERC20Response.decode(new minimal_1.default.Reader(data));
        });
    };
    MsgClientImpl.prototype.ConvertERC20ToCoin = function (request) {
        var data = exports.MsgConvertERC20ToCoin.encode(request).finish();
        var promise = this.rpc.request('kava.evmutil.v1beta1.Msg', 'ConvertERC20ToCoin', data);
        return promise.then(function (data) {
            return exports.MsgConvertERC20ToCoinResponse.decode(new minimal_1.default.Reader(data));
        });
    };
    return MsgClientImpl;
}());
exports.MsgClientImpl = MsgClientImpl;
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
