// @kava-labs/javascript-sdk/lib/proto/kava/hard/v1beta1/genesis.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisAccumulationTime = exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var hard_1 = require("../../../kava/hard/v1beta1/hard");
var timestamp_1 = require("../../../google/protobuf/timestamp");
var coin_1 = require("../../../cosmos/base/v1beta1/coin");
exports.protobufPackage = 'kava.hard.v1beta1';
function createBaseGenesisState() {
    return {
        params: undefined,
        previousAccumulationTimes: [],
        deposits: [],
        borrows: [],
        totalSupplied: [],
        totalBorrowed: [],
        totalReserves: [],
    };
}
exports.GenesisState = {
    encode: function (message, writer) {
        var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e, e_6, _f;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.params !== undefined) {
            hard_1.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        try {
            for (var _g = __values(message.previousAccumulationTimes), _h = _g.next(); !_h.done; _h = _g.next()) {
                var v = _h.value;
                exports.GenesisAccumulationTime.encode(v, writer.uint32(18).fork()).ldelim();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_h && !_h.done && (_a = _g.return)) _a.call(_g);
            }
            finally { if (e_1) throw e_1.error; }
        }
        try {
            for (var _j = __values(message.deposits), _k = _j.next(); !_k.done; _k = _j.next()) {
                var v = _k.value;
                hard_1.Deposit.encode(v, writer.uint32(26).fork()).ldelim();
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_k && !_k.done && (_b = _j.return)) _b.call(_j);
            }
            finally { if (e_2) throw e_2.error; }
        }
        try {
            for (var _l = __values(message.borrows), _m = _l.next(); !_m.done; _m = _l.next()) {
                var v = _m.value;
                hard_1.Borrow.encode(v, writer.uint32(34).fork()).ldelim();
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_m && !_m.done && (_c = _l.return)) _c.call(_l);
            }
            finally { if (e_3) throw e_3.error; }
        }
        try {
            for (var _o = __values(message.totalSupplied), _p = _o.next(); !_p.done; _p = _o.next()) {
                var v = _p.value;
                coin_1.Coin.encode(v, writer.uint32(42).fork()).ldelim();
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_p && !_p.done && (_d = _o.return)) _d.call(_o);
            }
            finally { if (e_4) throw e_4.error; }
        }
        try {
            for (var _q = __values(message.totalBorrowed), _r = _q.next(); !_r.done; _r = _q.next()) {
                var v = _r.value;
                coin_1.Coin.encode(v, writer.uint32(50).fork()).ldelim();
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_r && !_r.done && (_e = _q.return)) _e.call(_q);
            }
            finally { if (e_5) throw e_5.error; }
        }
        try {
            for (var _s = __values(message.totalReserves), _t = _s.next(); !_t.done; _t = _s.next()) {
                var v = _t.value;
                coin_1.Coin.encode(v, writer.uint32(58).fork()).ldelim();
            }
        }
        catch (e_6_1) { e_6 = { error: e_6_1 }; }
        finally {
            try {
                if (_t && !_t.done && (_f = _s.return)) _f.call(_s);
            }
            finally { if (e_6) throw e_6.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseGenesisState();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = hard_1.Params.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.previousAccumulationTimes.push(exports.GenesisAccumulationTime.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.deposits.push(hard_1.Deposit.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.borrows.push(hard_1.Borrow.decode(reader, reader.uint32()));
                    break;
                case 5:
                    message.totalSupplied.push(coin_1.Coin.decode(reader, reader.uint32()));
                    break;
                case 6:
                    message.totalBorrowed.push(coin_1.Coin.decode(reader, reader.uint32()));
                    break;
                case 7:
                    message.totalReserves.push(coin_1.Coin.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            params: isSet(object.params) ? hard_1.Params.fromJSON(object.params) : undefined,
            previousAccumulationTimes: Array.isArray(object === null || object === void 0 ? void 0 : object.previousAccumulationTimes)
                ? object.previousAccumulationTimes.map(function (e) {
                    return exports.GenesisAccumulationTime.fromJSON(e);
                })
                : [],
            deposits: Array.isArray(object === null || object === void 0 ? void 0 : object.deposits)
                ? object.deposits.map(function (e) { return hard_1.Deposit.fromJSON(e); })
                : [],
            borrows: Array.isArray(object === null || object === void 0 ? void 0 : object.borrows)
                ? object.borrows.map(function (e) { return hard_1.Borrow.fromJSON(e); })
                : [],
            totalSupplied: Array.isArray(object === null || object === void 0 ? void 0 : object.totalSupplied)
                ? object.totalSupplied.map(function (e) { return coin_1.Coin.fromJSON(e); })
                : [],
            totalBorrowed: Array.isArray(object === null || object === void 0 ? void 0 : object.totalBorrowed)
                ? object.totalBorrowed.map(function (e) { return coin_1.Coin.fromJSON(e); })
                : [],
            totalReserves: Array.isArray(object === null || object === void 0 ? void 0 : object.totalReserves)
                ? object.totalReserves.map(function (e) { return coin_1.Coin.fromJSON(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.params !== undefined &&
            (obj.params = message.params ? hard_1.Params.toJSON(message.params) : undefined);
        if (message.previousAccumulationTimes) {
            obj.previousAccumulationTimes = message.previousAccumulationTimes.map(function (e) { return (e ? exports.GenesisAccumulationTime.toJSON(e) : undefined); });
        }
        else {
            obj.previousAccumulationTimes = [];
        }
        if (message.deposits) {
            obj.deposits = message.deposits.map(function (e) {
                return e ? hard_1.Deposit.toJSON(e) : undefined;
            });
        }
        else {
            obj.deposits = [];
        }
        if (message.borrows) {
            obj.borrows = message.borrows.map(function (e) {
                return e ? hard_1.Borrow.toJSON(e) : undefined;
            });
        }
        else {
            obj.borrows = [];
        }
        if (message.totalSupplied) {
            obj.totalSupplied = message.totalSupplied.map(function (e) {
                return e ? coin_1.Coin.toJSON(e) : undefined;
            });
        }
        else {
            obj.totalSupplied = [];
        }
        if (message.totalBorrowed) {
            obj.totalBorrowed = message.totalBorrowed.map(function (e) {
                return e ? coin_1.Coin.toJSON(e) : undefined;
            });
        }
        else {
            obj.totalBorrowed = [];
        }
        if (message.totalReserves) {
            obj.totalReserves = message.totalReserves.map(function (e) {
                return e ? coin_1.Coin.toJSON(e) : undefined;
            });
        }
        else {
            obj.totalReserves = [];
        }
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f;
        var message = createBaseGenesisState();
        message.params =
            object.params !== undefined && object.params !== null
                ? hard_1.Params.fromPartial(object.params)
                : undefined;
        message.previousAccumulationTimes =
            ((_a = object.previousAccumulationTimes) === null || _a === void 0 ? void 0 : _a.map(function (e) {
                return exports.GenesisAccumulationTime.fromPartial(e);
            })) || [];
        message.deposits =
            ((_b = object.deposits) === null || _b === void 0 ? void 0 : _b.map(function (e) { return hard_1.Deposit.fromPartial(e); })) || [];
        message.borrows = ((_c = object.borrows) === null || _c === void 0 ? void 0 : _c.map(function (e) { return hard_1.Borrow.fromPartial(e); })) || [];
        message.totalSupplied =
            ((_d = object.totalSupplied) === null || _d === void 0 ? void 0 : _d.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        message.totalBorrowed =
            ((_e = object.totalBorrowed) === null || _e === void 0 ? void 0 : _e.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        message.totalReserves =
            ((_f = object.totalReserves) === null || _f === void 0 ? void 0 : _f.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        return message;
    },
};
function createBaseGenesisAccumulationTime() {
    return {
        collateralType: '',
        previousAccumulationTime: undefined,
        supplyInterestFactor: '',
        borrowInterestFactor: '',
    };
}
exports.GenesisAccumulationTime = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.collateralType !== '') {
            writer.uint32(10).string(message.collateralType);
        }
        if (message.previousAccumulationTime !== undefined) {
            timestamp_1.Timestamp.encode(toTimestamp(message.previousAccumulationTime), writer.uint32(18).fork()).ldelim();
        }
        if (message.supplyInterestFactor !== '') {
            writer.uint32(26).string(message.supplyInterestFactor);
        }
        if (message.borrowInterestFactor !== '') {
            writer.uint32(34).string(message.borrowInterestFactor);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseGenesisAccumulationTime();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.collateralType = reader.string();
                    break;
                case 2:
                    message.previousAccumulationTime = fromTimestamp(timestamp_1.Timestamp.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.supplyInterestFactor = reader.string();
                    break;
                case 4:
                    message.borrowInterestFactor = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            collateralType: isSet(object.collateralType)
                ? String(object.collateralType)
                : '',
            previousAccumulationTime: isSet(object.previousAccumulationTime)
                ? fromJsonTimestamp(object.previousAccumulationTime)
                : undefined,
            supplyInterestFactor: isSet(object.supplyInterestFactor)
                ? String(object.supplyInterestFactor)
                : '',
            borrowInterestFactor: isSet(object.borrowInterestFactor)
                ? String(object.borrowInterestFactor)
                : '',
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.collateralType !== undefined &&
            (obj.collateralType = message.collateralType);
        message.previousAccumulationTime !== undefined &&
            (obj.previousAccumulationTime =
                message.previousAccumulationTime.toISOString());
        message.supplyInterestFactor !== undefined &&
            (obj.supplyInterestFactor = message.supplyInterestFactor);
        message.borrowInterestFactor !== undefined &&
            (obj.borrowInterestFactor = message.borrowInterestFactor);
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d;
        var message = createBaseGenesisAccumulationTime();
        message.collateralType = (_a = object.collateralType) !== null && _a !== void 0 ? _a : '';
        message.previousAccumulationTime =
            (_b = object.previousAccumulationTime) !== null && _b !== void 0 ? _b : undefined;
        message.supplyInterestFactor = (_c = object.supplyInterestFactor) !== null && _c !== void 0 ? _c : '';
        message.borrowInterestFactor = (_d = object.borrowInterestFactor) !== null && _d !== void 0 ? _d : '';
        return message;
    },
};
function toTimestamp(date) {
    var seconds = numberToLong(date.getTime() / 1000);
    var nanos = (date.getTime() % 1000) * 1000000;
    return { seconds: seconds, nanos: nanos };
}
function fromTimestamp(t) {
    var millis = t.seconds.toNumber() * 1000;
    millis += t.nanos / 1000000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return o;
    }
    else if (typeof o === 'string') {
        return new Date(o);
    }
    else {
        return fromTimestamp(timestamp_1.Timestamp.fromJSON(o));
    }
}
function numberToLong(number) {
    return long_1.default.fromNumber(number);
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
