// @kava-labs/javascript-sdk/lib/proto/ethermint/feemarket/v1/feemarket.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = 'ethermint.feemarket.v1';
function createBaseParams() {
    return {
        noBaseFee: false,
        baseFeeChangeDenominator: 0,
        elasticityMultiplier: 0,
        initialBaseFee: long_1.default.ZERO,
        enableHeight: long_1.default.ZERO,
    };
}
exports.Params = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.noBaseFee === true) {
            writer.uint32(8).bool(message.noBaseFee);
        }
        if (message.baseFeeChangeDenominator !== 0) {
            writer.uint32(16).uint32(message.baseFeeChangeDenominator);
        }
        if (message.elasticityMultiplier !== 0) {
            writer.uint32(24).uint32(message.elasticityMultiplier);
        }
        if (!message.initialBaseFee.isZero()) {
            writer.uint32(32).int64(message.initialBaseFee);
        }
        if (!message.enableHeight.isZero()) {
            writer.uint32(40).int64(message.enableHeight);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseParams();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.noBaseFee = reader.bool();
                    break;
                case 2:
                    message.baseFeeChangeDenominator = reader.uint32();
                    break;
                case 3:
                    message.elasticityMultiplier = reader.uint32();
                    break;
                case 4:
                    message.initialBaseFee = reader.int64();
                    break;
                case 5:
                    message.enableHeight = reader.int64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            noBaseFee: isSet(object.noBaseFee) ? Boolean(object.noBaseFee) : false,
            baseFeeChangeDenominator: isSet(object.baseFeeChangeDenominator)
                ? Number(object.baseFeeChangeDenominator)
                : 0,
            elasticityMultiplier: isSet(object.elasticityMultiplier)
                ? Number(object.elasticityMultiplier)
                : 0,
            initialBaseFee: isSet(object.initialBaseFee)
                ? long_1.default.fromString(object.initialBaseFee)
                : long_1.default.ZERO,
            enableHeight: isSet(object.enableHeight)
                ? long_1.default.fromString(object.enableHeight)
                : long_1.default.ZERO,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.noBaseFee !== undefined && (obj.noBaseFee = message.noBaseFee);
        message.baseFeeChangeDenominator !== undefined &&
            (obj.baseFeeChangeDenominator = Math.round(message.baseFeeChangeDenominator));
        message.elasticityMultiplier !== undefined &&
            (obj.elasticityMultiplier = Math.round(message.elasticityMultiplier));
        message.initialBaseFee !== undefined &&
            (obj.initialBaseFee = (message.initialBaseFee || long_1.default.ZERO).toString());
        message.enableHeight !== undefined &&
            (obj.enableHeight = (message.enableHeight || long_1.default.ZERO).toString());
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c;
        var message = createBaseParams();
        message.noBaseFee = (_a = object.noBaseFee) !== null && _a !== void 0 ? _a : false;
        message.baseFeeChangeDenominator = (_b = object.baseFeeChangeDenominator) !== null && _b !== void 0 ? _b : 0;
        message.elasticityMultiplier = (_c = object.elasticityMultiplier) !== null && _c !== void 0 ? _c : 0;
        message.initialBaseFee =
            object.initialBaseFee !== undefined && object.initialBaseFee !== null
                ? long_1.default.fromValue(object.initialBaseFee)
                : long_1.default.ZERO;
        message.enableHeight =
            object.enableHeight !== undefined && object.enableHeight !== null
                ? long_1.default.fromValue(object.enableHeight)
                : long_1.default.ZERO;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
